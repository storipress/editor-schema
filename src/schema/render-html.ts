import type { DOMOutputSpec } from '@tiptap/pm/model'
import { htmlToOutputSpec } from '~/utils/convert'

export function renderHTML(html: string, allowIframe = false): DOMOutputSpec[] {
  if (!html) {
    return []
  }

  const content: any = h('div', {}, htmlToOutputSpec(html, allowIframe))
  const lastChild = content[content.length - 1]
  // if last child is a text node, remove it
  const hasEmptyChild = Array.isArray(lastChild) && (lastChild.length === 0 || lastChild[0].trim() === '')

  return [hasEmptyChild ? content.slice(0, -1) : content]
}

export function h(
  tag: string,
  attr: Record<string, string>,
  ...children: DOMOutputSpec[] | [DOMOutputSpec[]]
): DOMOutputSpec {
  if (isNestedOutputSpec(children)) {
    return [tag, attr || {}, ...children[0].filter(Boolean)]
  }

  return [tag, attr || {}, ...children.filter(Boolean)]
}

function isOutputSpecArray(x: unknown[]): x is DOMOutputSpec[] {
  if (Array.isArray(x[0]) && typeof x[0][0] === 'string') {
    return true
  }
  return false
}

function isNestedOutputSpec(x: unknown[]): x is [DOMOutputSpec[]] {
  if (!Array.isArray(x[0])) {
    return false
  }

  const first = x[0].filter(Boolean)
  return isOutputSpecArray(first)
}
