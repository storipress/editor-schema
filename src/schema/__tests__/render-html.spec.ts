import { html } from 'proper-tags'
import { h, renderHTML } from '../render-html'

describe('h', () => {
  it('render array children', () => {
    const res = h('div', {}, [null, h('span', {}, 'foo')] as any)

    expect(res).toMatchSnapshot()
  })

  it('render normal children', () => {
    const res = h('div', {}, h('span', {}, 'foo'))

    expect(res).toMatchSnapshot()
  })

  it('render with null children', () => {
    const res = h('div', {}, null as any, h('span', {}, 'foo'))

    expect(res).toMatchSnapshot()
  })
})

describe('renderHTML', () => {
  it('can handle plain text', () => {
    expect(renderHTML('foo')).toMatchInlineSnapshot(`
      [
        [
          "div",
          {},
          [
            "div",
            {},
            "foo",
          ],
        ],
      ]
    `)
  })

  it('can handle html', () => {
    expect(renderHTML(html`<p>foo <strong>bar</strong></p>`)).toMatchInlineSnapshot(`
      [
        [
          "div",
          {},
          [
            "p",
            {},
            "foo ",
            [
              "strong",
              {},
              "bar",
            ],
          ],
        ],
      ]
    `)
  })
})
