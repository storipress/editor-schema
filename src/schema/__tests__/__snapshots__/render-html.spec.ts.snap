// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`h > render array children 1`] = `
[
  "div",
  {},
  [
    "span",
    {},
    "foo",
  ],
]
`;

exports[`h > render normal children 1`] = `
[
  "div",
  {},
  [
    "span",
    {},
    "foo",
  ],
]
`;

exports[`h > render with null children 1`] = `
[
  "div",
  {},
  [
    "span",
    {},
    "foo",
  ],
]
`;
